Clarify cleanup in docs.

ekpyron opened the pull request:
Noted by @kuzdogan.

ekpyron committed:
Clarify cleanup in docs.

Co-authored-by: matheusaaguiar <95899911+matheusaaguiar@users.noreply.github.com>

kuzdogan commented on docs/internals/variable_cleanup.rst:
```diff
@@ -30,23 +30,24 @@ cleans input data when it is loaded onto the stack.
+|               |               |exception in the future  |
++---------------+---------------+-------------------------+
+|unsigned       |higher bits    |currently silently masks |
+|integers       |zeroed         |to to a valid value, i.e.|
```
```suggestion
|integers       |zeroed         |to a valid value, i.e.|
```

kuzdogan commented:
Mostly because I lack the context, the "valid values" of the last two are not clear for me. 

For signed integers what is word referring to? 2 bytes or 2x256 bits? 

For unsigned isn't it actually "left-padding" rather than masking? my understanding is for example the number `01010` will be `00000....00001010` in 256 bits.

ekpyron commented:
Ok, maybe we should rewrite this even more aggressively :-).

All values "live" in 256-bit types anyways (ultimately on the EVM level everything is 256-bits wide), so there is no "left-padding", but rather masking, i.e. the value already has higher-order bits, they are just set to zero.

But just to make absolutely clear that we're on the same page:

What the valid values are depends on the size of the type - so at the example of signed and unsigned 8-bit types:

Signed 8-bit Integers (``int8``)
---------------------------------------
For an 8-bit signed integer for example, the valid values are:

positive:
```
0000...0000 0000 0000
0000...0000 0000 0001
0000...0000 0000 0010
....
0000...0000 0111 1111
```
or negative:
```
1111...1111 1111 1111
1111...1111 1111 1110
....
1111...1111 1000 0000
```

Cleanup just means setting all the (256-8) higher (i.e. left) bits to the sign bit, i.e. the 8th bit from the right.

So the following are examples of invalid signed 8-bit integer values and the results of their cleanup:

Positive cases:
```
invalid value: 1111...1111 0000 0000
sign bit: 0
cleaned value: 0000...0000 0000 0000
```

```
invalid value: 0101...1101 0000 0000
sign bit: 0
cleaned value: 0000...0000 0000 0000
```

```
invalid value: 0101...1101 0001 0110
sign bit: 0
cleaned value: 0000...0000 0001 0110
```

Negative cases:
```
invalid value: 0000...0000 1111 1111
sign bit: 1
cleaned value: 1111...1111 1111 1111
```


```
invalid value: 0101...1101 1111 1111
sign bit: 1
cleaned value: 1111...1111 1111 1111
```

```
invalid value: 0101...1101 1001 1111
sign bit: 1
cleaned value: 1111...1111 1001 1111
```

Unsigned 8-bit Integers (``uint8``)
--------------------------------------------
For an 8-bit unsigned integer for example, the valid values are:

```
0000...0000 0000 0000
0000...0000 0000 0001
0000...0000 0000 0010
....
0000...0000 1111 1111
```

Cleanup just plainly means setting all the (256-8) higher (i.e.) left bits to zero (or equivalently: masking with ``0000...0000 1111 1111``).

So for example:

```
invalid value: 1111...1111 0000 0000
cleaned value: 0000...0000 0000 0000
```

```
invalid value: 0101...1101 0000 0000
cleaned value: 0000...0000 0000 0000
```

```
invalid value: 0101...1101 0001 0110
cleaned value: 0000...0000 0001 0110
```

```
invalid value: 0000...0001 1111 1111
cleaned value: 0000...0000 1111 1111
```

```
invalid value: 0101...1101 1111 1111
cleaned value: 0000...0000 1111 1111
```

```
invalid value: 0101...1101 1001 1111
cleaned value: 0000...0000 1001 1111
```



So I hope this ultimately clarifies, what the docs are **supposed** to say here :-).
If that's right, we can think about how to best phrase this, s.t. it actually makes that understandable :-).

ekpyron commented:
So in the 8-bit signed integer case, in ``sign-extended word``, ``word`` refers to the actually relevant 8 bits, i.e. the right-most 8 bits - if you consider those and sign-extend them to 256-bits you get all the valid 8-bit values embedded into 256-bits.

kuzdogan commented:
> All values "live" in 256-bit types anyways

This was the key and the examples given made it crystal clear, thanks!

In case of word, I'd consider changing the "word" (pun unintended) if it is not a used term in EVM or Solidity. I think something explicit like "sign-extended 256 bit number" is more clear.

leonardoalt commented:
What's the status here @ekpyron ?

ekpyron commented:
> What's the status here @ekpyron ?

Decide how best to rephrase or whether to just merge :-). To be fair, this is just 3 days old, though ;-).

leonardoalt commented:
3 days??? you mean `stale` :eyes:

ekpyron commented:
> 3 days??? you mean `stale` eyes

You can also just approve it and we merge it :-D. It's an improvement already, isn't it?

nikola-matic commented:
> Ok, maybe we should rewrite this even more aggressively :-).
> 
> All values "live" in 256-bit types anyways (ultimately on the EVM level everything is 256-bits wide), so there is no "left-padding", but rather masking, i.e. the value already has higher-order bits, they are just set to zero.
> 
> But just to make absolutely clear that we're on the same page:
> 
> What the valid values are depends on the size of the type - so at the example of signed and unsigned 8-bit types:
> ## Signed 8-bit Integers (`int8`)
> 
> For an 8-bit signed integer for example, the valid values are:
> 
> positive:
> 
> ```
> 0000...0000 0000 0000
> 0000...0000 0000 0001
> 0000...0000 0000 0010
> ....
> 0000...0000 0111 1111
> ```
> 
> or negative:
> 
> ```
> 1111...1111 1111 1111
> 1111...1111 1111 1110
> ....
> 1111...1111 1000 0000
> ```
> 
> Cleanup just means setting all the (256-8) higher (i.e. left) bits to the sign bit, i.e. the 8th bit from the right.
> 
> So the following are examples of invalid signed 8-bit integer values and the results of their cleanup:
> 
> Positive cases:
> 
> ```
> invalid value: 1111...1111 0000 0000
> sign bit: 0
> cleaned value: 0000...0000 0000 0000
> ```
> 
> ```
> invalid value: 0101...1101 0000 0000
> sign bit: 0
> cleaned value: 0000...0000 0000 0000
> ```
> 
> ```
> invalid value: 0101...1101 0001 0110
> sign bit: 0
> cleaned value: 0000...0000 0001 0110
> ```
> 
> Negative cases:
> 
> ```
> invalid value: 0000...0000 1111 1111
> sign bit: 1
> cleaned value: 1111...1111 1111 1111
> ```
> 
> ```
> invalid value: 0101...1101 1111 1111
> sign bit: 1
> cleaned value: 1111...1111 1111 1111
> ```
> 
> ```
> invalid value: 0101...1101 1001 1111
> sign bit: 1
> cleaned value: 1111...1111 1001 1111
> ```
> 
> ## Unsigned 8-bit Integers (`uint8`)
> 
> For an 8-bit unsigned integer for example, the valid values are:
> 
> ```
> 0000...0000 0000 0000
> 0000...0000 0000 0001
> 0000...0000 0000 0010
> ....
> 0000...0000 1111 1111
> ```
> 
> Cleanup just plainly means setting all the (256-8) higher (i.e.) left bits to zero (or equivalently: masking with `0000...0000 1111 1111`).
> 
> So for example:
> 
> ```
> invalid value: 1111...1111 0000 0000
> cleaned value: 0000...0000 0000 0000
> ```
> 
> ```
> invalid value: 0101...1101 0000 0000
> cleaned value: 0000...0000 0000 0000
> ```
> 
> ```
> invalid value: 0101...1101 0001 0110
> cleaned value: 0000...0000 0001 0110
> ```
> 
> ```
> invalid value: 0000...0001 1111 1111
> cleaned value: 0000...0000 1111 1111
> ```
> 
> ```
> invalid value: 0101...1101 1111 1111
> cleaned value: 0000...0000 1111 1111
> ```
> 
> ```
> invalid value: 0101...1101 1001 1111
> cleaned value: 0000...0000 1001 1111
> ```
> 
> So I hope this ultimately clarifies, what the docs are **supposed** to say here :-). If that's right, we can think about how to best phrase this, s.t. it actually makes that understandable :-).

To me, a layman, this comment does a better job of explaining variable cleanup than the contents of the docs :laughing: Perhaps adding a couple of example of signed and unsigned cleanup would make sense?

bshastry commented on docs/internals/variable_cleanup.rst:
```diff
@@ -30,23 +30,24 @@ cleans input data when it is loaded onto the stack.
++---------------+---------------+-------------------------+
+|unsigned       |higher bits    |currently silently masks |
+|integers       |zeroed         |to a valid value, i.e.   |
+|               |               |all higher bits are set  |
```
What does `higher bits zeroed` mean for unsigned integers wrt valid values? Aren't valid values [0, 2**n - 1]?

matheusaaguiar replied:
The table is describing the rule applied if the value is valid, i.e, fits the `n` bits length of the type. Currently, the same is done for invalid values also.
So, in cases where the type has less than 256 bits, the `higher = 256 - n` bits will be zeroed.

nishant-sachdeva replied:
Marking as resolved.

bshastry commented on docs/internals/variable_cleanup.rst:
```diff
@@ -30,23 +30,24 @@ cleans input data when it is loaded onto the stack.
++---------------+---------------+-------------------------+
+|bool           |0 or 1         |results in 1             |
++---------------+---------------+-------------------------+
+|signed integers|higher bits    |currently silently       |
```
Arent' valid values for signed integers `(-2**(n-1), 2**(n-1)]?

matheusaaguiar replied:
I think he meant that valid values have their `higher bits = 256 - n` set to the sign bit... which is the same for invalid values...

nishant-sachdeva replied:
Marking as resolved.

bshastry reviewed (commented):
Minor comment about wording of valid values for signed and unsigned integers, it is clear for enum and bool types.

matheusaaguiar commented on docs/internals/variable_cleanup.rst:
```diff
@@ -30,23 +30,24 @@ cleans input data when it is loaded onto the stack.
 
 Different types have different rules for cleaning up invalid values:
```
```suggestion
The following table describes the rules applied for valid and invalid values of different types, where ``higher bits`` refers to the remaining bits in case the type has less than 256 bits.
```

matheusaaguiar commented:
I'd suggest also to improve the section initial line from:
```
When a value is shorter than 256 bit, in some cases the remaining bits must be cleaned. 
```
to:
```
Ultimately, all values in the EVM are stored in 256 bit words. Thus, when the type of a value has less than 256 bits, in some cases the remaining bits must be cleaned.
```

ekpyron commented:
I took over @matheusaaguiar's suggestions.
We could of course additionally also add the contents of https://github.com/ethereum/solidity/pull/13381#issuecomment-1211800625 to the docs, if that helps...

matheusaaguiar commented:
I think that a small example might help people visualize what cleanup does. For me it's well explained, but I would say I'm a little biased because I worked with this in the recent overflow checks PR. So, it would be nice to have other opinions.

nikola-matic commented:
Rebased to get the build in order. @matheusaaguiar can you take a look and possibly make some changes if necessary so we can try to get this merged next week once @ekpyron comes back?

nishant-sachdeva reviewed (dismissed):
LGTM

matheusaaguiar commented:
> Rebased to get the build in order. @matheusaaguiar can you take a look and possibly make some changes if necessary so we can try to get this merged next week once @ekpyron comes back?

Sure!

ekpyron reviewed (commented):
I'm fine with the changes and merging this (but cannot approve, since I'm the original author and also probably not the right person to approve, since I'm too used to the cleanup rules myself).

kuzdogan commented on docs/internals/variable_cleanup.rst:
```diff
@@ -28,25 +29,45 @@ the boolean values before they are used as the condition for
+.. code-block:: none
+
+    0101...1101 0000 0000   invalid value
+    0000...0000 0000 0000   cleaned value
```
Maybe add signed integer examples to clarify the "signextend" term?

```suggestion
    0000...0000 0000 0000   cleaned value
    
For positive 8-bit signed integers:
.. code-block:: none
    Valid values:
    0000...0000 0000 0001
    0000...0000 0000 0010
    ....
    0000...0000 0111 1111

    invalid value: 0101...1101 0000 0100
    sign bit: 0
    cleaned value: 0000...0000 0000 0100

For negative 8-bit signed integers:
.. code-block:: none
    Valid values:
    1111...1111 1111 1111
    1111...1111 1111 1110
    ....
    1111...1111 1000 0000

    invalid value: 0000...0000 1111 1111
    sign bit: 1
    cleaned value: 1111...1111 1111 1111
```

matheusaaguiar replied:
Done.

matheusaaguiar reviewed (approved):
Approving and enabling auto-merge.
